### PR TITLE
Add prompt-lookup (n-gram) speculative decoding

### PR DESCRIPTION
**Pull request by fdagostino**
Implements the proposal in #425: **prompt-lookup decoding** (PLD / n-gram speculation) — draft-model-free speculative decoding, as requested for Python in ml-explore/mlx-lm#851 and implemented in llama.cpp and `transformers` (`prompt_lookup_num_tokens`).

### What

- `promptLookupDraft(history:maxNGramLength:minNGramLength:maxTokens:)`: pure n-gram lookup over the prompt+generated token history. Longest trailing pattern first, most recent occurrence wins, linear scan (an n-gram hash index à la llama.cpp is a possible follow-up if profiling warrants it).
- `PromptLookupTokenIterator`: mirrors `SpeculativeTokenIterator`'s verify/accept/rewind machinery with the lookup as draft source. Public init parameters: `numDraftTokens` (default 10), `maxNGramLength` (3), `minNGramLength` (1) — matching the reference implementation's defaults.

### Correctness at temperature > 0

The PLD proposal is deterministic — there is no draft distribution `q`. Accepting a proposed token iff it equals the token sampled from the target's own distribution keeps the output distribution exactly the target's at any temperature, so PLD needs none of the residual-sampling machinery that model-based drafting requires. Documented in the iterator's doc comment.

### Two hardenings over the existing speculative machinery

1. **Per-round rewind headroom check**: the round only runs if every cache layer is trimmable *and* — for window-bounded caches — has enough headroom that the verify pass cannot wrap the window mid-round. When the headroom is gone (e.g. a Gemma sliding-window layer wrapped), the iterator permanently falls back to single-token generation instead of corrupting the cache. This is the guard proposed as fix (a) in #424 — the silent-rollback failure mode reported there cannot occur here by construction.
2. **Lookup miss ⇒ plain single step**, not a wasted batched verify: PLD costs nothing on non-repetitive stretches.

### Testing

- 6 deterministic unit tests for the lookup (longest-pattern preference, most-recent occurrence, clamping, degenerate histories).
- Iterator contract tests against a deterministic transition model (same contract as the existing speculative-decoding test model, plus a forward-call counter):
  - **Exact token-stream parity** with `TokenIterator` on cyclic text — with strictly fewer model forwards and `acceptedDraftTokenCount > 0`.
  - **Parity preserved when a rotating cache wraps mid-generation** (the fallback path).
- `xcodebuild build-for-testing` + `MLXLMTests` via `xcrun xctest` (same as CI): 193 tests pass on this branch.

### Follow-ups (not in this PR, happy to do either)

- `ChatSession` integration (a config mirroring `SpeculativeDecodingConfig`).
- N-gram hash index for long contexts.

### CODE REVIEW

File: Libraries/MLXLMCommon/PromptLookupDecoding.swift [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+        switch try model.prepare(input, cache: cache, state: state, windowSize: windowSize) {
+        case .tokens(let tokens):
+            y = tokens
+        case .logits(let result):
+            var logits = result.logits[0..., -1, 0...]
+            logits = processor?.process(logits: logits) ?? logits
+            let token = sampler.sample(logits: logits)
+            processor?.didSample(token: token)
+            y = .init(tokens: token)
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/PromptLookupDecoding.swift:185]: Does this need the:

```swift
        pendingTokens.append(tokenValue)
        history.append(tokenValue)
```

like `singleStep()`.

Comment (fdagostino) [reply Libraries/MLXLMCommon/PromptLookupDecoding.swift:192]: Yes — good catch, it did. Fixed in 7947445.

The `.tokens` case hands the prompt back for the iterator to forward, so nothing is generated yet. The `.logits` case is different: `prepare` already evaluated the final prompt position, so the token sampled from those logits **is** the first generated token. Assigning it to `y` alone carried it into the next forward pass but never emitted it — and never entered the n-gram corpus the lookup draws proposals from, so it was invisible to matching too.

The regression test uses a model whose `prepare` returns `.logits`. Without the fix it reproduces cleanly — prompt lookup starts one position ahead of plain generation:

```
PLD   → [4, 5, 6, 7, 0, 1, ...]
plain → [3, 4, 5, 6, 7, 0, ...]
```

This is the same prefill-bonus handling `singleStep()` and `MTPSpeculativeTokenIterator` already do.

I also rebased onto current `main` and adapted three signatures that moved since this branch was written (cfadbb3): `TokenIteratorProtocol.state` is now a public requirement, `newCache` throws, and `prepare` takes `PrefillParameters` instead of a window size. No behavior change — the test model ignores the prefill parameters, as the other iterator test models do.

Full suite green locally (618 XCTest + 749 swift-testing), docs verified.
